File: generate-icons.js
#!/usr/bin/env node

/**
 * HexSapper Icon Generator
 * Generates PNG icons for PWA using Node.js canvas
 * 
 * Install dependencies first:
 * npm install canvas
 * 
 * Then run:
 * node generate-icons.js
 */

const fs = require('fs');
const path = require('path');

// Check if canvas is available
let createCanvas;
try {
    const { createCanvas: cc } = require('canvas');
    createCanvas = cc;
} catch (e) {
    console.log('Canvas module not available. Please run: npm install canvas');
    console.log('Or use generate-icons.html in a browser instead.');
    process.exit(1);
}

const sizes = [72, 96, 128, 144, 152, 192, 384, 512];
const iconsDir = path.join(__dirname, 'icons');

function drawIcon(size) {
    const canvas = createCanvas(size, size);
    const ctx = canvas.getContext('2d');
    const s = size;
    
    // Background gradient
    const gradient = ctx.createLinearGradient(0, 0, s, s);
    gradient.addColorStop(0, '#1a1a2e');
    gradient.addColorStop(1, '#16213e');
    ctx.fillStyle = gradient;
    
    // Rounded rectangle
    const r = s * 0.12;
    ctx.beginPath();
    ctx.roundRect(0, 0, s, s, r);
    ctx.fill();
    
    // Hexagon
    const cx = s / 2;
    const cy = s / 2;
    const hexR = s * 0.25;
    
    ctx.beginPath();
    for (let i = 0; i < 6; i++) {
        const angle = (Math.PI / 3) * i - Math.PI / 6;
        const x = cx + hexR * Math.cos(angle);
        const y = cy + hexR * Math.sin(angle);
        if (i === 0) ctx.moveTo(x, y);
        else ctx.lineTo(x, y);
    }
    ctx.closePath();
    
    const hexGrad = ctx.createLinearGradient(cx, cy - hexR, cx, cy + hexR);
    hexGrad.addColorStop(0, '#4a4a6a');
    hexGrad.addColorStop(1, '#2a2a3a');
    ctx.fillStyle = hexGrad;
    ctx.fill();
    ctx.strokeStyle = '#e94560';
    ctx.lineWidth = Math.max(2, s * 0.01);
    ctx.stroke();
    
    // Mine circle
    ctx.beginPath();
    ctx.arc(cx, cy, hexR * 0.35, 0, Math.PI * 2);
    ctx.fillStyle = '#e94560';
    ctx.fill();
    
    // Mine spikes
    ctx.strokeStyle = '#e94560';
    ctx.lineWidth = Math.max(3, s * 0.015);
    ctx.lineCap = 'round';
    const spikeLen = hexR * 0.5;
    const innerLen = hexR * 0.25;
    
    for (let i = 0; i < 8; i++) {
        const angle = (Math.PI / 4) * i;
        ctx.beginPath();
        ctx.moveTo(cx + innerLen * Math.cos(angle), cy + innerLen * Math.sin(angle));
        ctx.lineTo(cx + spikeLen * Math.cos(angle), cy + spikeLen * Math.sin(angle));
        ctx.stroke();
    }
    
    return canvas;
}

// Ensure icons directory exists
if (!fs.existsSync(iconsDir)) {
    fs.mkdirSync(iconsDir, { recursive: true });
}

// Generate all icons
console.log('Generating icons...\n');

sizes.forEach(size => {
    const canvas = drawIcon(size);
    const filename = path.join(iconsDir, `icon-${size}.png`);
    
    const buffer = canvas.toBuffer('image/png');
    fs.writeFileSync(filename, buffer);
    
    console.log(`✓ Created icon-${size}.png (${size}x${size})`);
});

console.log('\n✅ All icons generated successfully!');
console.log(`📁 Icons saved to: ${iconsDir}/`);

File: generate-icons.py
#!/usr/bin/env python3
"""
HexSapper Icon Generator
Generates simple PNG icons for PWA without external dependencies.
Uses pure Python to create valid PNG files.
"""

import struct
import zlib
import os

def create_png(width, height, pixels):
    """Create a PNG file from pixel data."""
    
    def png_chunk(chunk_type, data):
        chunk_len = struct.pack('>I', len(data))
        chunk_crc = struct.pack('>I', zlib.crc32(chunk_type + data) & 0xffffffff)
        return chunk_len + chunk_type + data + chunk_crc
    
    # PNG signature
    signature = b'\x89PNG\r\n\x1a\n'
    
    # IHDR chunk
    ihdr_data = struct.pack('>IIBBBBB', width, height, 8, 2, 0, 0, 0)
    ihdr = png_chunk(b'IHDR', ihdr_data)
    
    # IDAT chunk (image data)
    raw_data = b''
    for y in range(height):
        raw_data += b'\x00'  # Filter type: None
        for x in range(width):
            r, g, b = pixels[x, y]
            raw_data += bytes([r, g, b])
    
    compressed = zlib.compress(raw_data, 9)
    idat = png_chunk(b'IDAT', compressed)
    
    # IEND chunk
    iend = png_chunk(b'IEND', b'')
    
    return signature + ihdr + idat + iend

def create_icon(size):
    """Create a HexSapper icon."""
    pixels = {}
    
    s = size
    cx, cy = s // 2, s // 2
    
    for x in range(s):
        for y in range(s):
            # Background gradient
            bg_ratio = (x + y) / (2 * s)
            r = int(26 + (22 - 26) * bg_ratio)
            g = int(26 + (33 - 26) * bg_ratio)
            b = int(46 + (62 - 46) * bg_ratio)
            
            # Check if inside rounded rectangle
            corner_radius = int(s * 0.12)
            in_corner = False
            
            # Check corners
            corners = [
                (corner_radius, corner_radius),
                (s - corner_radius, corner_radius),
                (corner_radius, s - corner_radius),
                (s - corner_radius, s - corner_radius)
            ]
            
            for ccx, ccy in corners:
                dx = x - ccx if x < ccx else (x - (s - ccx) if x > s - ccx else 0)
                dy = y - ccy if y < ccy else (y - (s - ccy) if y > s - ccy else 0)
                if dx != 0 or dy != 0:
                    if (x < ccx or x > s - ccx) and (y < ccy or y > s - ccy):
                        dist = ((x - ccx) ** 2 + (y - ccy) ** 2) ** 0.5
                        if dist > corner_radius:
                            in_corner = True
                            break
            
            if in_corner:
                pixels[x, y] = (26, 26, 46)
                continue
            
            # Hexagon
            hex_r = int(s * 0.25)
            hex_points = []
            for i in range(6):
                angle = (3.14159 / 3) * i - 3.14159 / 6
                hx = cx + int(hex_r * (0.866 if i % 2 else 1) * (1 if i < 3 else -1 if i > 3 else 0))
                hy = cy + int(hex_r * (1 if i % 2 == 0 else 0.866) * (-1 if i < 2 else 1 if i > 3 else 0))
                hex_points.append((hx, hy))
            
            # Simple hexagon approximation
            dx = abs(x - cx)
            dy = abs(y - cy)
            hex_dist = dx * 0.866 + dy
            
            in_hex = hex_dist < hex_r * 0.95
            
            if in_hex:
                # Hexagon gradient
                hex_ratio = y / s
                r = int(74 + (42 - 74) * hex_ratio)
                g = int(74 + (58 - 74) * hex_ratio)
                b = int(106 + (58 - 106) * hex_ratio)
                
                # Mine circle
                mine_r = int(hex_r * 0.35)
                dist_to_center = ((x - cx) ** 2 + (y - cy) ** 2) ** 0.5
                
                if dist_to_center < mine_r:
                    r, g, b = 233, 69, 96  # Mine circle color
                elif dist_to_center < mine_r * 1.5:
                    # Mine spikes (simplified)
                    angle = ((x - cx) ** 2 + (y - cy) ** 2) ** 0.5
                    if angle > 0:
                        spike_angle = ((x - cx) / angle) * 57.3
                        if abs(spike_angle % 45) < 10 or abs(spike_angle % 45) > 35:
                            r, g, b = 233, 69, 96
                
                pixels[x, y] = (r, g, b)
            else:
                pixels[x, y] = (r, g, b)
    
    return create_png(size, size, pixels)

def main():
    sizes = [72, 96, 128, 144, 152, 192, 384, 512]
    icons_dir = os.path.join(os.path.dirname(__file__), 'icons')
    
    os.makedirs(icons_dir, exist_ok=True)
    
    print('Generating icons...\n')
    
    for size in sizes:
        png_data = create_icon(size)
        filename = os.path.join(icons_dir, f'icon-{size}.png')
        
        with open(filename, 'wb') as f:
            f.write(png_data)
        
        print(f'✓ Created icon-{size}.png ({size}x{size})')
    
    print('\n✅ All icons generated successfully!')
    print(f'📁 Icons saved to: {icons_dir}/')

if __name__ == '__main__':
    main()
